### PR TITLE
reconciles machine deletion hooks

### PR DESCRIPTION
**Pull request by p0lyn0mial**

### CODE REVIEW

File: pkg/operator/machinedeletionhooks/machinedeletionhooks.go
Comment (p0lyn0mial) [inline pkg/operator/machinedeletionhooks/machinedeletionhooks.go:94]: @deads2k @hasbro17 how about this? does it also make sense?

File: pkg/operator/machinedeletionhooks/machinedeletionhooks.go
Comment (p0lyn0mial) [inline pkg/operator/machinedeletionhooks/machinedeletionhooks.go:106]: this is the current plan

File: pkg/operator/machinedeletionhooks/machinedeletionhooks.go
Comment (deads2k) [reply pkg/operator/machinedeletionhooks/machinedeletionhooks.go:102]: > @deads2k @hasbro17 how about this? does it also make sense?

This would be represented as a check in clustermembercontroller to avoid adding members that correspond to machines that are pending deletion?

**Comment by deads2k** [comment]
Is this still a WIP?

File: pkg/operator/machinedeletionhooks/machinedeletionhooks.go
Comment (deads2k) [inline pkg/operator/machinedeletionhooks/machinedeletionhooks.go:86]: What is the disadvantage of simply adding the hook to every master machine?  that would effectively eliminate one class of races since the deletion hook would be present long before etcd was present on the machine.

Comment (p0lyn0mial) [reply pkg/operator/machinedeletionhooks/machinedeletionhooks.go:94]: It won't solve the issue. In the case that a Machine is deleted before the member is promoted, the etcd operator is expected to not promote the new member, and remove the deletion hook to allow the Machine to be removed from the cluster.

File: pkg/operator/machinedeletionhooks/machinedeletionhooks.go
Comment (p0lyn0mial) [inline pkg/operator/machinedeletionhooks/machinedeletionhooks.go:163]: do we want to record an event?

**Comment by p0lyn0mial** [comment]
> Is this still a WIP?

it is ready for review, I am working on adding more unit tests.

**Comment by p0lyn0mial** [comment]
/hold

until we update the member removal controller.

**Comment by p0lyn0mial** [comment]
@hasbro17 @deads2k  a friendly ping to review this pr

File: pkg/operator/machinedeletionhooks/machinedeletionhooks.go
Comment (hasbro17) [inline pkg/operator/machinedeletionhooks/machinedeletionhooks.go:72]: Are we only re-syncing on `kube-system` configmaps?
I would have thought this controller would need to react to changes on Machines for hook changes, and potentially nodes or etcd pods to detect member addition/removal.

File: pkg/operator/machinedeletionhooks/machinedeletionhooks.go
Comment (hasbro17) [inline pkg/operator/machinedeletionhooks/machinedeletionhooks.go:142]: Style nit: You don't have to change it :) but I find these blocks more readable when we can return/continue early to reduce nesting if possible.
```suggestion
		if currentMemberIPListSet.Has(memberMachinePendingDeletionIP) {
			// Ignore machines that still have a member present
			continue
		}
		memberMachinePendingDeletionCopy := memberMachinePendingDeletion.DeepCopy()
		removeMachineDeletionHook(memberMachinePendingDeletionCopy)
		if _, err := c.machineClient.Update(ctx, memberMachinePendingDeletionCopy, metav1.UpdateOptions{}); err == nil {
			klog.Infof("successfully removed the deletion hook from machine %v", memberMachinePendingDeletion.Name)
		}
		return err
	}
```

File: pkg/operator/machinedeletionhooks/machinedeletionhooks.go
Comment (hasbro17) [reply pkg/operator/machinedeletionhooks/machinedeletionhooks.go:139]: I think yes it's probably worth noting when we unblock a machine.

**Review by hasbro17 — COMMENTED** [review]
Still going through the unit tests but the implementation looks good to me. Just a question on the informers used to sync this controller.

File: pkg/operator/machinedeletionhooks/machinedeletionhooks.go
Comment (p0lyn0mial) [reply pkg/operator/machinedeletionhooks/machinedeletionhooks.go:142]: np

File: pkg/operator/machinedeletionhooks/machinedeletionhooks.go
Comment (p0lyn0mial) [reply pkg/operator/machinedeletionhooks/machinedeletionhooks.go:72]: good point, added `masterMachineInformer` and changed the resync time to 1 minute

**Comment by p0lyn0mial** [comment]
applied the review comments, ptal

File: pkg/operator/machinedeletionhooks/machinedeletionhooks.go
Comment (deads2k) [inline pkg/operator/machinedeletionhooks/machinedeletionhooks.go:164]: this is just `machineMemberCopy`, now right?

File: pkg/operator/machinedeletionhooks/machinedeletionhooks.go
Comment (deads2k) [inline pkg/operator/machinedeletionhooks/machinedeletionhooks.go:86]: an error here will stop you from removing your machine deletion hook.  you sure you want to combine them?

File: pkg/operator/machinedeletionhooks/machinedeletionhooks.go
Comment (deads2k) [inline pkg/operator/machinedeletionhooks/machinedeletionhooks.go:125]: `currentVotingMemberIPListSet`?

File: pkg/operator/machinedeletionhooks/machinedeletionhooks.go
Comment (deads2k) [inline pkg/operator/machinedeletionhooks/machinedeletionhooks.go:144]: doesn't have a voting member.

File: pkg/operator/ceohelpers/common.go
Comment (deads2k) [inline pkg/operator/ceohelpers/common.go:60]: spewdump member in an error looks too long to me.  How about an ID or an IP or something short.

**Comment by deads2k** [comment]
this looks plausible to me.

File: pkg/operator/machinedeletionhooks/machinedeletionhooks.go
Comment (p0lyn0mial) [reply pkg/operator/machinedeletionhooks/machinedeletionhooks.go:164]: yes, will update

File: pkg/operator/machinedeletionhooks/machinedeletionhooks.go
Comment (p0lyn0mial) [reply pkg/operator/machinedeletionhooks/machinedeletionhooks.go:125]: nope, the list holds both voting members and learners.
we also need to remove hooks from machines that host "non-promotable" learners.

File: pkg/operator/ceohelpers/common.go
Comment (p0lyn0mial) [reply pkg/operator/ceohelpers/common.go]: done

File: pkg/operator/machinedeletionhooks/machinedeletionhooks.go
Comment (p0lyn0mial) [reply pkg/operator/machinedeletionhooks/machinedeletionhooks.go:78]: updated

File: pkg/operator/machinedeletionhooks/machinedeletionhooks.go
Comment (hasbro17) [inline pkg/operator/machinedeletionhooks/machinedeletionhooks.go:102]: Why does the hooks controller care about whether we're over/under the desired control-plane replica count?
I mentioned this in https://github.com/openshift/cluster-etcd-operator/pull/768#discussion_r836926279 but it's the member removal controller that needs to make this consideration when deciding whether it's safe to remove/scale-down a voting member. Plus the member removal controller will also take care of cleaning up learner members pending deletion.

So for the hooks controller, all that would be required is to remove the deletion hook from any machine that isn't mapped to an etcd member (learner or voting).

**Review by hasbro17 — CHANGES_REQUESTED** [review]

File: pkg/operator/machinedeletionhooks/machinedeletionhooks.go
Comment (hasbro17) [reply pkg/operator/machinedeletionhooks/machinedeletionhooks.go:102]: Apologies if I missed this before on my first pass but this just stood out to me when reviewing #768

**Review by hasbro17 — APPROVED** [review]
/approve

File: pkg/operator/machinedeletionhooks/machinedeletionhooks.go
Comment (p0lyn0mial) [reply pkg/operator/machinedeletionhooks/machinedeletionhooks.go:102]: kind of, all master machines are machines that host etcd members but yeah, np, I can rename it.

**Comment by p0lyn0mial** [comment]
we are going to wire the controller in https://github.com/openshift/cluster-etcd-operator/pull/768

**Comment by deads2k** [comment]
where does this use ReadDesiredControlPlaneReplicasCount  from https://github.com/openshift/cluster-etcd-operator/pull/767/files ?  I wouldn't expect that value to be pertinent.

**Comment by p0lyn0mial** [comment]
> where does this use ReadDesiredControlPlaneReplicasCount from https://github.com/openshift/cluster-etcd-operator/pull/767/files ?

it doesn't anymore. I removed the check that required it.

**Comment by p0lyn0mial** [comment]
oh, it looks like https://github.com/openshift/cluster-etcd-operator/pull/768 got first.

768 contained commits introduced in this PR.

**Pull request closed without merging**
